### PR TITLE
RM#71890 BILLOFMATERIAL import main panel and results

### PR DESCRIPTION
**Pull request by dar-axelor**

### CODE REVIEW

File: axelor-production/src/main/java/com/axelor/apps/production/web/BillOfMaterialImportController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        Beans.get(BillOfMaterialImportService.class);
+
+    ImportHistory importHistory = billOfMaterialImportService.processImport(billOfMaterialImport);
+
+    File readFile = MetaFiles.getPath(importHistory.getLogMetaFile()).toFile();
+    response.setNotify(
+        FileUtils.readFileToString(readFile, StandardCharsets.UTF_8)
+            .replaceAll("(\r\n|\n\r|\r|\n)", "<br />"));
+
+    billOfMaterialImportService.setStatusToImported(billOfMaterialImport);
```
Comment (cce-axelor) [inline axelor-production/src/main/java/com/axelor/apps/production/web/BillOfMaterialImportController.java:55]: why is this action outside of the processImport method ?

File: axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImporter.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        billOfMaterialImport.addBillOfMaterialImportLineListItem(billOfMaterialImportLine);
+        billOfMaterialImportLine.setDescription(billOfMaterialImport.getDocumentMetaFile());
+        billOfMaterialImportLine.setBillOfMaterialImport(billOfMaterialImport);
+      }
+    } catch (Exception e) {
+      listener.handle(billOfMaterialImportLine, e);
+    }
+  }
+
+  @Transactional
```
Comment (cce-axelor) [inline axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImporter.java:165]: @Transactional(rollbackOn = {Exception.class})

File: axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImporter.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        setBoMLevel(listener, billOfMaterialImportLine);
+        if (i % FETCH_LIMIT == 0) {
+          JPA.clear();
+        }
+        i++;
+      }
+    }
+  }
+
+  @Transactional
```
Comment (cce-axelor) [inline axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImporter.java:148]: @Transactional(rollbackOn = {Exception.class})

File: axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImportServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    }
+    if (billOfMaterialImportLineOp.isEmpty()) {
+      throw new AxelorException(
+          billOfMaterialImport,
+          TraceBackRepository.CATEGORY_MISSING_FIELD,
+          ProductionExceptionMessage.BOM_IMPORT_NO_MAIN_BILL_OF_MATERIAL_GENERATED);
+    }
+    BillOfMaterialImportLine billOfMaterialImportLine = billOfMaterialImportLineOp.get();
+
+    return generateBillOfMaterialFromImportLine(billOfMaterialImportLine).orElse(null);
```
Comment (cce-axelor) [inline axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImportServiceImpl.java:122]: can be done in one line
something like billOfMaterialImportLineOp.map(boImportLine -> generateBillOfMaterialFromImportLine(boImportLine)).orElse()

File: axelor-production/src/main/java/com/axelor/apps/production/db/repo/BillOfMaterialManagementRepository.java [hunk 5]
```diff
@@ -45,7 +49,7 @@ public BillOfMaterial save(BillOfMaterial billOfMaterial) {
       billOfMaterial.setFullName(billOfMaterial.getName());
     }
 
-    return super.save(billOfMaterial);
+    return billOfMaterial;
```
[the remark this answers is not in the document]
Comment (dar-axelor) [reply axelor-production/src/main/java/com/axelor/apps/production/db/repo/BillOfMaterialManagementRepository.java:52]: Saving first sets an ID to the BOM which is used to compute the name.

**Pull request merged**